# Exercícios Práticos: Consumindo uma API REST com Python

## API escolhida: DummyJSON

Neste notebook, você irá praticar o consumo de uma API pública chamada **DummyJSON**.

Ela fornece dados fictícios em formato JSON, como produtos, usuários, carrinhos, posts e tarefas. É uma API muito útil para treinar consumo de API REST, análise de JSON e transformação de dados com Python.

### Documentação oficial

Acesse a documentação da API:

https://dummyjson.com/docs

Documentação específica de produtos:

https://dummyjson.com/docs/products

---

## Objetivos dos exercícios

Ao final deste notebook, você deverá ser capaz de:

- Fazer requisições HTTP usando Python;
- Consumir endpoints com o método `GET`;
- Enviar dados usando `POST`;
- Atualizar dados usando `PUT` ou `PATCH`;
- Simular exclusão usando `DELETE`;
- Interpretar respostas em JSON;
- Transformar dados da API em tabelas com Pandas;
- Criar pequenas análises a partir dos dados recebidos.

---

## Importante

A DummyJSON é uma API de testes. As operações de criação, atualização e exclusão são **simuladas**.

Isso significa que a API responde como se tivesse criado, atualizado ou excluído um recurso, mas os dados não são realmente persistidos no servidor.

---

## Como usar este notebook

Leia cada enunciado com atenção e complete os trechos marcados com `TODO`.

Não apague as células de enunciado.

Quando terminar, execute todas as células na ordem.

# Parte 1 — Preparação do ambiente

Antes de consumir uma API, precisamos importar algumas bibliotecas.

Usaremos:

- `requests`: para fazer chamadas HTTP;
- `pandas`: para transformar dados JSON em tabelas;
- `json`: para exibir algumas respostas de forma mais organizada.

In [ ]:
import requests
import pandas as pd
import json

Vamos definir a URL base da API.

Sempre que formos chamar um endpoint da DummyJSON, começaremos por essa URL.

In [ ]:
url_base = "https://dummyjson.com"

# Parte 2 — Primeiro contato com a API

## Exercício 1 — Testar se a API está funcionando

Faça uma requisição `GET` para o endpoint abaixo:

```text
https://dummyjson.com/test
```

Depois, exiba:

- O código de status HTTP;
- A resposta em JSON.

### Dica

Use:

```python
requests.get(...)
```

In [ ]:
response = requests.get(f'{url_base}/test')

In [ ]:
response.status_code

200

In [ ]:
response.json()

{'status': 'ok', 'method': 'GET'}

## Perguntas do Exercício 1

Responda abaixo, em texto:

1. Qual foi o status code retornado?
2. O que esse status code significa?
3. A API respondeu em qual formato de dados?

**Respostas:**

1. 200
2. conexão estabelecida
3. json

# Parte 3 — Consumindo produtos com GET

A partir de agora, trabalharemos com o recurso `products`.

Um recurso, em uma API REST, representa uma entidade do sistema.

Neste caso:

```text
/products
```

representa a coleção de produtos da API.

## Exercício 2 — Buscar todos os produtos

Faça uma requisição `GET` para:

```text
https://dummyjson.com/products
```

Depois, exiba:

- O status code;
- As chaves principais do JSON retornado;
- A quantidade de produtos retornados nesta primeira página;
- O total de produtos disponíveis na API.

### Observação

A resposta da API não é uma lista diretamente. Ela retorna um objeto JSON com algumas chaves, como:

- `products`
- `total`
- `skip`
- `limit`

In [ ]:
# Exercício 2

# TODO: faça a requisição GET

# TODO: converta a resposta para JSON

# TODO: exiba o status code

# TODO: exiba as chaves principais do JSON

# TODO: exiba a quantidade de produtos retornados nesta chamada

# TODO: exiba o total de produtos disponíveis na API


In [ ]:
f'{url_base}/produts'

'https://dummyjson.com/produts'

In [ ]:
response1 = requests.get(f'{url_base}/products')

In [ ]:
response1.status_code

200

In [ ]:
response1.json()

In [ ]:
chaves = response1.json().keys()
chaves

dict_keys(['products', 'total', 'skip', 'limit'])

In [ ]:
quantidade = len(response1.json()['products'])
quantidade

30

In [ ]:
total = response1.json()['total']

In [ ]:
total

194

## Exercício 3 — Visualizar o primeiro produto

Usando os dados do exercício anterior, exiba apenas o primeiro produto da lista.

Depois, responda:

1. Qual é o nome do produto?
2. Qual é o preço?
3. Qual é a categoria?
4. Qual é a avaliação média, campo `rating`?

In [ ]:
produto = 1

In [ ]:
response2 = requests.get(f'{url_base}/products/{produto}')

In [ ]:
response2.status_code

200

In [ ]:
response2.json()

**Respostas:**

1. Nome do produto: Essence Mascara Lash Princess
2. Preço: 9.99
3. Categoria: beauty
4. Rating: 2.56

# Parte 4 — Transformando JSON em tabela

Em projetos de dados e IA, consumir uma API normalmente é apenas o começo.

Depois de receber os dados em JSON, é comum transformar esses dados em uma tabela para análise.

## Exercício 4 — Criar um DataFrame de produtos

Transforme a lista de produtos em um DataFrame do Pandas.

Depois, exiba as 5 primeiras linhas.

In [ ]:
# Exercício 4

# TODO: transforme dados["products"] em um DataFrame
df_produtos = None

# TODO: exiba as 5 primeiras linhas


In [ ]:
f'{url_base}/products'

'https://dummyjson.com/products'

In [ ]:
response3 = requests.get(f'{url_base}/products')

In [ ]:
response3.status_code

200

In [ ]:
response3.json()

In [ ]:
products = pd.DataFrame(response3.json())

In [ ]:
products

In [ ]:
json_1 = response3.json()

In [ ]:
df_produtos = pd.json_normalize(json_1['products'])

In [ ]:
df_produtos.head(5)

## Exercício 5 — Inspecionar o DataFrame

Agora que você tem uma tabela, descubra:

1. Quantas linhas e colunas existem?
2. Quais são os nomes das colunas?
3. Quais colunas parecem numéricas?
4. Quais colunas parecem textuais?

In [ ]:
# Exercício 5

# TODO: exiba o formato do DataFrame com .shape


# TODO: exiba as colunas do DataFrame


# TODO: exiba informações gerais com .info()


In [ ]:
df_produtos.shape

(30, 27)

In [ ]:
coluna = df_produtos.columns

In [ ]:
len(coluna)

27

In [ ]:
len( df_produtos.select_dtypes(include=['object']).columns)

17

**Respostas:**

1. Linhas e colunas: 30 linhas, 27 colunas
2. Colunas existentes: 27
3. Colunas numéricas: 10
4. Colunas textuais: 17

# Parte 5 — GET com parâmetros

Muitas APIs permitem enviar parâmetros na URL.

Exemplo:

```text
https://dummyjson.com/products?limit=10&skip=0
```

Nessa URL:

- `limit=10` indica que queremos no máximo 10 produtos;
- `skip=0` indica que não queremos pular nenhum produto.

Em Python, podemos enviar esses parâmetros usando o argumento `params` da biblioteca `requests`.

## Exercício 6 — Buscar apenas 10 produtos

Faça uma requisição `GET` para `/products`, mas retornando apenas 10 produtos.

Use `params` em vez de escrever os parâmetros direto na URL.

In [ ]:
# Exercício 6

url = BASE_URL + "/products"

params = {
    # TODO: adicione o parâmetro limit com valor 10
}

# TODO: faça a requisição GET com params=params
resposta = None

# TODO: converta para JSON
dados_10 = None

# TODO: exiba a quantidade de produtos retornados


# TODO: crie um DataFrame com esses 10 produtos
df_10 = None

# TODO: exiba o DataFrame


In [ ]:
f'{url_base}/products'

params = {
    'limit': 10
}

In [ ]:
response4 = requests.get(f'{url_base}/products',params=params)

In [ ]:
response4.status_code

200

In [ ]:
dados_10 = response4.json()

In [ ]:
quantidade_produto = len(dados_10['products'])

In [ ]:
quantidade_produto

10

In [ ]:
df_10 = pd.json_normalize(dados_10['products'])

In [ ]:
df_10

## Exercício 7 — Paginação com skip

Agora busque os próximos 10 produtos.

Use:

- `limit=10`
- `skip=10`

Depois, compare os IDs dos produtos do Exercício 6 com os IDs deste exercício.

A ideia é perceber que a API está retornando uma segunda página de resultados.

In [ ]:
# Exercício 7

url = BASE_URL + "/products"

params = {
    # TODO: defina limit igual a 10
    # TODO: defina skip igual a 10
}

# TODO: faça a requisição GET
resposta = None

# TODO: converta a resposta para JSON
dados_pagina_2 = None

# TODO: crie um DataFrame
df_pagina_2 = None

# TODO: exiba as colunas id, title e price


In [ ]:
f'{url_base}/products'

'https://dummyjson.com/products'

In [ ]:
params = {
    'limit': 10,
    'skip': 10
}

In [ ]:
response5 = requests.get(f'{url_base}/products', params=params)

In [ ]:
response5.status_code

200

In [ ]:
dados_11 = response5.json()

In [ ]:
df_pagina2 = pd.json_normalize(dados_11['products'])

In [ ]:
df_pagina2[['id', 'title', 'price']]

**Pergunta:**

Qual foi a diferença entre `limit` e `skip`?
  

**Resposta:**
  limit diz quntos intens você quer ver.
  skip a partir de qual item deve começar a aparecer.


# Parte 6 — Selecionando campos específicos

Às vezes, uma API retorna muitos campos, mas você precisa apenas de alguns.

A DummyJSON permite usar o parâmetro `select` para escolher quais campos devem retornar.

Exemplo:

```text
select=title,price,category
```

## Exercício 8 — Buscar produtos com campos selecionados

Faça uma requisição para buscar 15 produtos retornando apenas os campos:

- `title`
- `price`
- `category`
- `rating`

Depois, transforme o resultado em DataFrame.

In [ ]:
# Exercício 8

url = BASE_URL + "/products"

params = {
    # TODO: limit igual a 15
    # TODO: select com os campos title,price,category,rating
}

# TODO: faça a requisição
resposta = None

# TODO: converta para JSON
dados_selecionados = None

# TODO: crie o DataFrame
df_selecionados = None

# TODO: exiba o DataFrame


In [ ]:
f'{url_base}/products'

'https://dummyjson.com/products'

In [ ]:
params = {
    'limit': 15,
    'select': {'title', 'price', 'category', 'rating'}
}

In [ ]:
response6 = requests.get(f'{url_base}/products', params=params)

In [ ]:
response6.status_code

200

In [ ]:
dados_15 = response6.json()

In [ ]:
df_selecionados = pd.json_normalize(dados_15['products'])

In [ ]:
df_selecionados

# Parte 7 — Busca textual na API

Algumas APIs permitem busca por texto.

Na DummyJSON, podemos pesquisar produtos usando:

```text
/products/search?q=phone
```

O parâmetro `q` representa o termo pesquisado.

## Exercício 9 — Pesquisar produtos por palavra-chave

Faça uma busca por produtos relacionados à palavra:

```text
phone
```

Depois:

1. Exiba a quantidade de resultados encontrados;
2. Crie um DataFrame;
3. Exiba as colunas `id`, `title`, `category`, `price` e `rating`.

In [ ]:
# Exercício 9

url = BASE_URL + "/products/search"

params = {
    # TODO: q igual a "phone"
}

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
dados_busca = None

# TODO: exiba o total de resultados


# TODO: crie o DataFrame
df_busca = None

# TODO: exiba apenas as colunas solicitadas


In [ ]:
f'{url_base}/products/search'

'https://dummyjson.com/products/search'

In [ ]:
params = {
    'q': 'phone'
}

In [ ]:
response7 = requests.get(f'{url_base}/products/search', params=params)

In [ ]:
response6.status_code

200

In [ ]:
dados_busca = response7.json()

In [ ]:
dados_busca

In [ ]:
df_dados_busca = pd.json_normalize(dados_busca['products'])

In [ ]:
df_dados_busca

## Exercício 10 — Repetir a busca com outro termo

Agora escolha outro termo de busca.

Sugestões:

- `laptop`
- `watch`
- `bag`
- `shoes`
- `perfume`

Depois, compare os resultados com a busca anterior.

In [ ]:
# Exercício 10

termo = ""  # TODO: escolha um termo de busca

url = BASE_URL + "/products/search"

params = {
    "q": termo
}

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
dados_busca_2 = None

# TODO: crie um DataFrame
df_busca_2 = None

# TODO: exiba id, title, category, price e rating


In [ ]:
f'{url_base}/products/search'

'https://dummyjson.com/products/search'

In [ ]:
params = {
    'q': 'bag'
}

In [ ]:
response8 = requests.get(f'{url_base}/products/search', params=params)

In [ ]:
response.status_code

200

In [ ]:
dados_busca_2 = response8.json()

In [ ]:
df_20 = pd.json_normalize(dados_busca_2['products'])

In [ ]:
df_20[['id', 'title', 'category', 'price', 'rating']]

**Perguntas:**

1. Qual termo você pesquisou?
2. Quantos resultados foram encontrados?
3. Os produtos encontrados fazem sentido para o termo pesquisado?
4. Que problema poderia acontecer se o usuário digitasse uma palavra errada?

**Respostas:**

1. BAG
2. 4
3. Sim
4. Ia dar ERROR

# Parte 8 — Categorias

Muitas APIs organizam recursos por categoria.

Na DummyJSON, podemos listar categorias de produtos usando:

```text
/products/category-list
```

E buscar produtos de uma categoria específica usando:

```text
/products/category/NOME_DA_CATEGORIA
```

## Exercício 11 — Listar categorias

Faça uma requisição para listar todas as categorias de produtos.

Depois, exiba a lista de categorias.

In [ ]:
# Exercício 11

url = BASE_URL + "/products/category-list"

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
categorias = None

# TODO: exiba as categorias


In [ ]:
f'{url_base}/products/category-list'

'https://dummyjson.com/products/category-list'

In [ ]:
response9 = requests.get(f'{url_base}/products/category-list')

In [ ]:
response9.status_code

200

In [ ]:
categorias = response9.json()

In [ ]:
df_categorias = pd.DataFrame(categorias)

In [ ]:
df_categorias

## Exercício 12 — Buscar produtos de uma categoria

Escolha uma categoria da lista anterior.

Depois, busque os produtos dessa categoria.

Exemplo de endpoint:

```text
/products/category/smartphones
```

In [ ]:
# Exercício 12

categoria_escolhida = ""  # TODO: escolha uma categoria

url = BASE_URL + f"/products/category/{categoria_escolhida}"

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
dados_categoria = None

# TODO: crie um DataFrame com os produtos da categoria
df_categoria = None

# TODO: exiba id, title, category, price, rating e stock


In [ ]:
categoria_escolhida = 'motorcycle'

In [ ]:
f'{url_base}/products/category/{categoria_escolhida}'

'https://dummyjson.com/products/category/motorcycle'

In [ ]:
response10 = requests.get(f'{url_base}/products/category/{categoria_escolhida}')

In [ ]:
response10.status_code

200

In [ ]:
uma_categoria = response10.json()

In [ ]:
df_categoria = pd.json_normalize(uma_categoria['products'])

In [ ]:
df_categoria

**Perguntas:**

1. Qual categoria você escolheu?
2. Quantos produtos vieram nessa categoria?
3. Qual foi o produto mais caro?
4. Qual foi o produto com maior avaliação?

**Respostas:**

1. motorcycle
2. 5
3. 14999.99
4. The Generic Motorcycle

# Parte 9 — Análise de dados com Pandas

Agora vamos usar os dados da API para fazer pequenas análises.

Essa etapa aproxima o consumo de API do trabalho de Ciência de Dados e IA.

## Exercício 13 — Estatísticas dos preços

Usando o DataFrame `df_categoria` ou outro DataFrame de produtos, calcule:

1. Preço médio;
2. Menor preço;
3. Maior preço;
4. Mediana dos preços.

In [ ]:
# Exercício 13

# TODO: calcule o preço médio


# TODO: calcule o menor preço


# TODO: calcule o maior preço


# TODO: calcule a mediana dos preços


In [ ]:
df_categoria['price'].mean()

np.float64(7699.99)

In [ ]:
df_categoria['price'].min()

2999.99

In [ ]:
df_categoria['price'].max()

14999.99

In [ ]:
df_categoria['price'].median()

7499.99

## Exercício 14 — Produtos mais bem avaliados

Ordene os produtos por `rating`, do maior para o menor.

Depois, exiba os 5 produtos com maior avaliação.

In [ ]:
# Exercício 14

# TODO: ordene o DataFrame pela coluna rating em ordem decrescente
df_ordenado_rating = None

# TODO: exiba os 5 primeiros produtos com as colunas title, category, price e rating


In [ ]:
df_categoria['rating'].sort_values(ascending=False)

In [ ]:
df_categoria[['title', 'title', 'category', 'price', 'rating']].head(5)

## Exercício 15 — Produtos com estoque baixo

Filtre os produtos com estoque menor ou igual a 10.

Depois, exiba:

- Nome do produto;
- Categoria;
- Preço;
- Estoque.

In [ ]:
# Exercício 15

# TODO: filtre produtos com stock <= 10
df_estoque_baixo = None

# TODO: exiba as colunas solicitadas


In [ ]:
df_estoque_baixo = df_categoria[df_categoria['stock'] <= 10 ]

In [ ]:
df_estoque_baixo

# Parte 10 — Criando dados com POST

O método `POST` é usado, em geral, para criar um novo recurso.

Na DummyJSON, a criação é simulada.

Endpoint:

```text
POST https://dummyjson.com/products/add
```

Você deverá enviar um corpo JSON com os dados do produto.

## Exercício 16 — Criar um produto fictício

Crie um produto fictício relacionado a tecnologia ou inteligência artificial.

O produto deve ter pelo menos:

- `title`
- `description`
- `price`
- `brand`
- `category`

Depois, envie esse produto usando `POST`.

In [ ]:
# Exercício 16

url = BASE_URL + "/products/add"

novo_produto = {
    # TODO: preencha os dados do produto
}

# TODO: envie o produto usando requests.post(..., json=novo_produto)
resposta = None

# TODO: exiba o status code


# TODO: exiba a resposta em JSON


In [ ]:
f'{url_base}/products/add'

'https://dummyjson.com/products/add'

In [ ]:
novo_produto = {
    'title': 'Coleira inteligente',
    'description': 'guie suas vacas em direção a novas seções de um pasto e as mantenham dentro de limites definidos.',
    'price': 9.90,
    'brad': 'Halter',
    'category': 'AgTech'

}

In [ ]:
response11 = requests.post(f'{url_base}/products/add', json=novo_produto)

In [ ]:
response11.status_code

201

In [ ]:
response11.json()

{'id': 195,
 'title': 'Coleira inteligente',
 'price': 9.9,
 'description': 'guie suas vacas em direção a novas seções de um pasto e as mantenham dentro de limites definidos.',
 'category': 'AgTech'}

**Perguntas:**

1. Qual status code foi retornado?
2. A API retornou um ID para o produto criado?
3. Esse produto foi realmente salvo no servidor? Explique com suas palavras.

**Respostas:**

1. 201
2. 1995
3. Sim. Como foi criado, automaticamente ele vai ao servidor.

# Parte 11 — Atualizando dados com PUT

O método `PUT` é usado para atualizar um recurso existente.

Na prática, muitas APIs usam `PUT` para atualizar um recurso inteiro e `PATCH` para atualizar apenas alguns campos.

Na DummyJSON, a atualização também é simulada.

Endpoint de exemplo:

```text
PUT https://dummyjson.com/products/1
```

## Exercício 17 — Atualizar o título de um produto

Escolha um produto existente, por exemplo o produto de ID 1.

Atualize o título dele para um nome criado por você.

Depois, exiba a resposta da API.

In [ ]:
# Exercício 17

id_produto = 1
url = BASE_URL + f"/products/{id_produto}"

dados_atualizados = {
    # TODO: informe um novo title
}

# TODO: envie a atualização com PUT
resposta = None

# TODO: exiba o status code


# TODO: exiba a resposta em JSON


In [ ]:
id_produto = 1

In [ ]:
f'{url_base}/products/{id_produto}'

'https://dummyjson.com/products/1'

In [ ]:
dados_atualizados = {
    'title': 'Kitkat'
}

In [ ]:
resposta = requests.put(f'{url_base}/products/{id_produto}', json=dados_atualizados)

In [ ]:
resposta.status_code

200

In [ ]:
resposta.json()

## Exercício 18 — Atualizar preço e estoque

Agora atualize dois campos do produto:

- `price`
- `stock`

Use o método `PUT` ou `PATCH`.

In [ ]:
# Exercício 18

id_produto = 1
url = BASE_URL + f"/products/{id_produto}"

dados_atualizados = {
    # TODO: informe price e stock
}

# TODO: use requests.put ou requests.patch
resposta = None

# TODO: exiba a resposta em JSON


In [ ]:
id_produto = 1

f'{url_base}/products/{id_produto}'

'https://dummyjson.com/products/1'

In [ ]:
dados_atualizados = {
    'price': 1.99,
    'stock': 250
}

In [ ]:
resposta = requests.patch(f'{url_base}/products/{id_produto}', json=dados_atualizados)

In [ ]:
resposta.status_code

200

In [ ]:
resposta.json()

**Perguntas:**

1. Qual método você usou: PUT ou PATCH?
2. Quais campos você alterou?
3. A API retornou os dados modificados?
4. Qual é a diferença conceitual entre atualizar dados e criar dados?

**Respostas:**

1. PATCH
2. PRICE, STOCK
3. Sim.
4. Atualizar: vai alterar um dado já existente, criar: Vai criar um novo dado.

# Parte 12 — Excluindo dados com DELETE

O método `DELETE` é usado para excluir um recurso.

Na DummyJSON, a exclusão é simulada.

Endpoint de exemplo:

```text
DELETE https://dummyjson.com/products/1
```

## Exercício 19 — Simular exclusão de um produto

Faça uma requisição `DELETE` para excluir o produto de ID 1.

Depois, verifique se a resposta possui os campos:

- `isDeleted`
- `deletedOn`

In [ ]:
# Exercício 19

id_produto = 1
url = BASE_URL + f"/products/{id_produto}"

# TODO: faça a requisição DELETE
resposta = None

# TODO: converta para JSON
dados_delete = None

# TODO: exiba o status code


# TODO: exiba a resposta em JSON


# TODO: verifique se isDeleted é True


In [ ]:
id_produto = 1

f'{url_base}/products/{id_produto}'

'https://dummyjson.com/products/1'

In [ ]:
resposta = requests.delete(f'{url_base}/products/{id_produto}')

In [ ]:
resposta.status_code

200

In [ ]:
dados_delete = resposta.json()

In [ ]:
dados_delete

In [ ]:
if not dados_delete['isDeleted']:
  print('O item não foi deletado')

**Perguntas:**

1. O campo `isDeleted` apareceu?
2. O campo `deletedOn` apareceu?
3. Em uma API real, por que o método DELETE precisa ser usado com cuidado?
4. Que tipo de permissão uma API real provavelmente exigiria antes de permitir exclusões?

**Respostas:**

1. Sim.
2. não
3. Porque uma vez deletado, a possibilidade de recuperar é quase nenhuma.
4.

# Parte 13 — Tratamento básico de erros

Nem toda requisição a uma API dá certo.

Podem ocorrer problemas como:

- Endpoint errado;
- Produto inexistente;
- Falta de internet;
- Falta de autorização;
- Erro interno no servidor.

Por isso, é importante verificar o `status_code` antes de confiar nos dados.

## Exercício 20 — Testar um produto inexistente

Faça uma requisição para buscar um produto com ID muito alto, por exemplo:

```text
/products/999999
```

Depois, exiba:

- O status code;
- A resposta em JSON;
- Uma mensagem amigável dizendo se a requisição deu certo ou não.

In [ ]:
# Exercício 20

url = BASE_URL + "/products/999999"

# TODO: faça a requisição GET
resposta = None

# TODO: exiba o status code


# TODO: se o status code for 200, exiba "Produto encontrado"
# TODO: caso contrário, exiba "Produto não encontrado ou erro na requisição"


# TODO: exiba a resposta JSON


In [ ]:
f'{url_base}/products/9999'

'https://dummyjson.com/products/9999'

In [ ]:
resposta = requests.get(f'{url_base}/products/9999')

In [ ]:
resposta = resposta.status_code

In [ ]:
if resposta == 200:
  print(f'Produto encontrado')
else:
  print(f'Produto não encontrado ou erro na requisição. ERRO: {resposta}')

Produto não encontrado ou erro na requisição. ERRO: 404


# Parte 14 — Criando uma função para consumir a API

Em vez de repetir código várias vezes, podemos criar funções.

Isso deixa o código mais limpo e reutilizável.

## Exercício 21 — Criar uma função para buscar produto por ID

Crie uma função chamada `buscar_produto_por_id`.

A função deve receber um ID e retornar os dados do produto.

Requisitos:

- Receber o ID como parâmetro;
- Fazer a chamada para `/products/{id}`;
- Verificar se o status code é 200;
- Retornar o JSON se der certo;
- Retornar `None` se não encontrar o produto.

In [ ]:
# Exercício 21

def buscar_produto_por_id(id_produto):
    # TODO: monte a URL

    # TODO: faça a requisição GET

    # TODO: verifique o status code

    # TODO: retorne o JSON se encontrar ou None caso contrário
    pass


# TODO: teste a função com um produto existente


# TODO: teste a função com um produto inexistente


In [ ]:
def buscar_produto_por_id(id_produto):
  f'{url_base}/products/{id_produto}'

  response = requests.get(f'{url_base}/products/{id_produto}')

  busca = response.status_code

  response.json()

  if busca == 200:
    print(f'{response.json()}')

  else:
    print(f'Não encontrado - {busca}')


In [ ]:
buscar_produto_por_id(1)

{'id': 1, 'title': 'Essence Mascara Lash Princess', 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.', 'category': 'beauty', 'price': 9.99, 'discountPercentage': 10.48, 'rating': 2.56, 'stock': 99, 'tags': ['beauty', 'mascara'], 'brand': 'Essence', 'sku': 'BEA-ESS-ESS-001', 'weight': 4, 'dimensions': {'width': 15.14, 'height': 13.08, 'depth': 22.99}, 'warrantyInformation': '1 week warranty', 'shippingInformation': 'Ships in 3-5 business days', 'availabilityStatus': 'In Stock', 'reviews': [{'rating': 3, 'comment': 'Would not recommend!', 'date': '2025-04-30T09:41:02.053Z', 'reviewerName': 'Eleanor Collins', 'reviewerEmail': 'eleanor.collins@x.dummyjson.com'}, {'rating': 4, 'comment': 'Very satisfied!', 'date': '2025-04-30T09:41:02.053Z', 'reviewerName': 'Lucas Gordon', 'reviewerEmail': 'lucas.gordon@x.dummyjson.com'}, {'rating': 5, 'comment': '

In [ ]:
buscar_produto_por_id(99999)

Não encontrado - 404


# Parte 15 — Mini desafio final

Agora você deverá juntar tudo o que praticou.

## Cenário

Imagine que você foi contratado para iniciar a análise de produtos de uma loja virtual.

Seu trabalho é consumir a API da DummyJSON, buscar produtos, transformar em tabela e gerar alguns insights iniciais.

## Exercício 22 — Relatório simples de produtos

Faça as seguintes etapas:

1. Busque todos os produtos disponíveis usando `limit=0`;
2. Transforme os produtos em um DataFrame;
3. Exiba a quantidade total de produtos;
4. Exiba a quantidade de categorias diferentes;
5. Mostre as 10 categorias com mais produtos;
6. Mostre os 10 produtos mais caros;
7. Mostre os 10 produtos com maior avaliação;
8. Calcule o preço médio por categoria;
9. Escolha uma categoria e explique brevemente o que você percebeu nos dados.

### Dica

Para buscar todos os produtos, use:

```python
params = {"limit": 0}
```

In [ ]:
# Exercício 22

url = BASE_URL + "/products"

params = {
    # TODO: use limit igual a 0
}

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
dados_todos = None

# TODO: crie um DataFrame com todos os produtos
df_todos = None

# TODO: quantidade total de produtos


# TODO: quantidade de categorias diferentes


# TODO: 10 categorias com mais produtos


# TODO: 10 produtos mais caros


# TODO: 10 produtos com maior avaliação


# TODO: preço médio por categoria


## Conclusão do mini desafio

Escreva um pequeno parágrafo respondendo:

1. Que tipo de dado essa API fornece?
2. Como esses dados poderiam ser usados em um projeto de IA?
3. Que cuidado você teria antes de usar dados de uma API real?
4. Qual foi a parte mais difícil do exercício?

**Resposta:**



# Desafio extra — Escolha outro recurso da API

A DummyJSON também possui outros recursos além de produtos.

Exemplos:

- `/users`
- `/posts`
- `/comments`
- `/carts`
- `/todos`
- `/recipes`

Escolha um desses recursos na documentação e crie uma nova análise.

## O que entregar

1. Endpoint escolhido;
2. Explicação do que esse endpoint retorna;
3. Requisição GET funcionando;
4. DataFrame criado com os dados;
5. Pelo menos 3 perguntas respondidas com análise dos dados.

Exemplo de perguntas:

- Quantos registros existem?
- Quais são os campos disponíveis?
- Existe algum campo numérico para análise?
- É possível agrupar por alguma categoria?
- Esses dados poderiam alimentar algum modelo de IA?

In [ ]:
# Desafio extra

# TODO: escolha outro recurso da API
endpoint = ""  # exemplo: "/users"

url = BASE_URL + endpoint

# TODO: faça a requisição GET
resposta = None

# TODO: converta para JSON
dados_extra = None

# TODO: investigue a estrutura do JSON


# TODO: crie um DataFrame


# TODO: faça suas análises


# Checklist de entrega

Antes de finalizar, confira se você conseguiu:

- [ ] Fazer uma requisição GET simples;
- [ ] Interpretar um JSON;
- [ ] Criar um DataFrame com dados da API;
- [ ] Usar parâmetros na URL;
- [ ] Fazer busca textual;
- [ ] Buscar produtos por categoria;
- [ ] Simular um POST;
- [ ] Simular um PUT ou PATCH;
- [ ] Simular um DELETE;
- [ ] Tratar um caso de erro;
- [ ] Criar uma função para buscar dados;
- [ ] Gerar uma pequena análise com Pandas;
- [ ] Explicar como o consumo de API pode ser útil em projetos de dados e IA.